ext/StaticArraysExt/utils.jl

```diff
@@ -14,4 +14,44 @@ Base.@propagate_inbounds function _promote_output_matrix(::Type{<:SMatrix}, mat_
 end
 Base.@propagate_inbounds function _promote_output_matrix(::Type{<:SMatrix}, mat_out, out_dim::Tuple)
     return SMatrix{out_dim[1],out_dim[2]}(mat_out)
-end
\ No newline at end of file
+end
+Base.@propagate_inbounds function _promote_output_matrix(::Type{T1}, ::Type{T2}, mat_out) where {T1<:SMatrix, T2<:AbstractMatrix}
+    return SMatrix{size(mat_out,1), size(mat_out,2)}(mat_out)
+end
+Base.@propagate_inbounds function _promote_output_vector(::Type{T1}, ::Type{T2}, vec_out) where {T1<:SVector, T2<:AbstractVector}
+    return SVector{length(vec_out)}(vec_out)
+end
+
+abstract type ArrayTrait end
+
+struct DenseArrayTrait <: ArrayTrait end
+struct StaticArrayTrait <: ArrayTrait end
+
+array_trait(::Type{<:Array}) = DenseArrayTrait()
+array_trait(::Type{<:SArray}) = StaticArrayTrait()
+array_trait(::Type{<:UnionAll}) = StaticArrayTrait()
+
+function _infer_types(::DenseArrayTrait, nmodes, T = Float64)
+    disp_type = Vector{T}
+    transform_type = Matrix{T}
+    return disp_type, transform_type
+end
+function _infer_types(::StaticArrayTrait, nmodes, T = Float64)
+    disp_type = SArray{Tuple{2*nmodes}, T}
+    transform_type = SArray{Tuple{2*nmodes, 2*nmodes}, T}
+    return disp_type, transform_type
+end
+function _infer_types(T1, T2, basis)
+    nmodes = basis.nmodes
+    elT1 = eltype(T1)
+    elT2 = eltype(T2)
+    disp_type1, _ = _infer_types(array_trait(T1), nmodes, elT1)
+    _, transform_type2 = _infer_types(array_trait(T2), nmodes, elT2)
+    return disp_type1, transform_type2
+end
+function _infer_types(T, basis)
+    nmodes = basis.nmodes
+    elT = eltype(T)
+    disp_type, transform_type = _infer_types(array_trait(T), nmodes, elT)
+    return disp_type, transform_type
+end
```

Comment:
```suggestion
function _infer_types(::Type{T1}, ::Type{T2}, basis) where {T1<:SVector, T2<:SMatrix}
    nmodes = basis.nmodes
    return T1{2*nmodes}, T2{2*nmodes,2*nmodes}
end
function _infer_types(::Type{T1}, ::Type{T2}, basis) where {T1<:SVector{<:Int}, T2<:SMatrix{<:Int, <:Int}}
    nmodes = basis.nmodes
    return T1, T2
end
function _infer_types(::Type{T}, basis) where {T<:SArray}
    nmodes = basis.nmodes
    return SVector{2*nmodes}, SMatrix{2*nmodes,2*nmodes}
end
```
Hi Feroz, I'm curious on whether or not you think this approach might be simpler. I'm not opposed to trait-based dispatch, but it might be too fancy for what we have here. To deal with any array other than a static array, we could also define a default function in `src/utils.jl` with something like
```
_infer_type(T1, T2, basis) = T1
_infer_type(T, basis) = T
```

Reply:
Hi, Andrew!

There was an error occurred following the type-based dispatch due to an attempt to` iterate` over the abstract type `SArray` itself, rather than an instance of a concrete subtype (e.g., `SVector, SMatrix`). 

The `iterate` function, designed for iterating over collections, is not defined for type objects. Consequently, the call` randchannel(SArray, qpairbasis)` led to the error as the` _infer_types` function, within `randchannel`, attempted to process the abstract `SArray `type, triggering the iterate method.

This following error is now resolved by enhancing the `_infer_types` functions to handle concrete `SArray` subtypes effectively.


```julia
 MethodError: no method matching iterate(::Type{SArray})
  The function iterate exists, but no method is defined for this combination of argument types.
  
  Closest candidates are:
    iterate(::Pipe, ::Int64)
     @ Base stream.jl:1297
    iterate(::Pipe)
     @ Base stream.jl:1296
    iterate(::Core.MethodMatch, ::Int64)
     @ Base deprecated.jl:293
    ...
  
  Stacktrace:
    [1] indexed_iterate(I::Type, i::Int64)
      @ Base ./tuple.jl:162
    [2] randchannel(::Type{SArray}, basis::QuadPairBasis{Int64}; ħ::Int64)
      @ Gabs ~/Desktop/Gabs/Gabfix/10/Gabs.jl/src/randoms.jl:96
    [3] randchannel(::Type{SArray}, basis::QuadPairBasis{Int64})
      @ Gabs ~/Desktop/Gabs/Gabfix/10/Gabs.jl/src/randoms.jl:95
    [4] macro expansion
      @ REPL[15]:22 [inlined]
    [5] macro expansion
      @ ~/.julia/juliaup/julia-1.11.3+0.x64.linux.gnu/share/julia/stdlib/v1.11/Test/src/Test.jl:1704 [inlined]                                          
    [6] top-level scope
      @ REPL[15]:2
    [7] eval
      @ ./boot.jl:430 [inlined]
    [8] eval_user_input(ast::Any, backend::REPL.REPLBackend, mod::Module)
      @ REPL ~/.julia/juliaup/julia-1.11.3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:245                                                
    [9] repl_backend_loop(backend::REPL.REPLBackend, get_module::Function)
      @ REPL ~/.julia/juliaup/julia-1.11.3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:342                                                
   [10] start_repl_backend(backend::REPL.REPLBackend, consumer::Any; get_module::Function)                                                              
      @ REPL ~/.julia/juliaup/julia-1.11.3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:327                                                
   [11] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool, backend::Any)                                                   
      @ REPL ~/.julia/juliaup/julia-1.11.3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:483                                                
   [12] run_repl(repl::REPL.AbstractREPL, consumer::Any)
      @ REPL ~/.julia/juliaup/julia-1.11.3+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:469                                                
   [13] (::Base.var"#1150#1152"{Bool, Symbol, Bool})(REPL::Module)
      @ Base ./client.jl:446
   [14] #invokelatest#2
      @ ./essentials.jl:1055 [inlined]
   [15] invokelatest
      @ ./essentials.jl:1052 [inlined]
   [16] run_main_repl(interactive::Bool, quiet::Bool, banner::Symbol, history_file::Bool, color_set::Bool)                                              
      @ Base ./client.jl:430
   [17] repl_main
      @ ./client.jl:567 [inlined]
   [18] _start()
      @ Base ./client.jl:541
```

What's left is to resolve `MethodError: no method matching iterate(::Type{Array})` since we want to avoid breaking changes. I will do this soon as well :)

Reply:
> What's left is to resolve MethodError: no method matching iterate(::Type{Array})

This is resolved! 

The existing functionality that does not break as requested also includes the following (`test_unitaries.jl)`:  
- `displace(Vector, Matrix, ...)`
- `displace(Array, ...)`

Some tests actually test passing the following convention of types:

- `op(Vector, Matrix,...)`
- `op(Array, ...)`

Hence, it is made sure that this convention also does not break.